README.md
# pymssql-utils
_pymssql-utils_ is a small library that wraps
[pymssql](https://github.com/pymssql/pymssql) to make your life easier.
It provides you with a higher-level API as well as various utility functions
for generating dynamic queries.

This module features:
* Higher-level API that reduces the amount of boilerplate required.
* Baked-in sensible defaults and usage patterns.
* Provides optional execution batching, similar to
  [_pyodbc's_](https://github.com/mkleehammer/pyodbc) `fast_executemany`.
* Provides consistent parsing between SQL Types and native Python types over different platforms and drivers.
* Makes it easy to serialize your data with
  [_orjson_](https://github.com/ijl/orjson).
* Provides you with simple and clear options for error handling.
* Extra utility functions for building dynamic SQL queries.
* Fixing various edge case bugs that arise when using _pymssql_.
* Fully type hinted.

This module's enforced opinions (check these work for you):
* Each execution opens and closes a connection using _pymssql_'s
  context management.
* Automatically converts certain data types for ease of use, e.g. `Decimal` -> `float`, `UUID` -> `str`.
  
When you shouldn't use this module:
* If you need fine-grained control over your cursors.
  
Please raise any suggestions or issues via GitHub.

## Status

This library is tested and stable. There will not be any breaking changes to the 
public API without a major version release.  There is also scope for expanding the 
library if new features are requested.

## Changes

See the repository's [GitHub releases](https://github.com/invokermain/pymssql-utils/releases)
or the `CHANGELOG.md`.

## Usage
### Installation

This library can be installed via pip: `pip install --upgrade pymssql-utils`.
This library requires `Python >= 3.7`.

If you want to serialize your results to JSON you can install the optional dependency `ORJSON`
by running `pip install --upgrade pymssql-utils[json]`.

If you want to cast your results to DataFrame you can install the optional dependency `Pandas`
by running `pip install --upgrade pymssql-utils[pandas]`.

### Quickstart

For querying the database this library provides two high-level methods:
 * `query`: executes a SQL operation that fetches the result, and DOES NOT commit the transaction (by default).
 * `execute`: executes a SQL operation that does not fetch the result (by default), and DOES commit the transaction.

This separation of _pymssql's_ `execute` is to make your code more explicit and readable.

Here is an example for running a simple query and accessing the returned data:
```python
>>> import pymssqlutils as sql
>>> result = sql.query(
      "SELECT SYSDATETIMEOFFSET() as now",
      server="..."
    )
>>> result.data
[{'now': datetime.datetime(2021, 1, 21, 23, 31, 11, 272299, tzinfo=datetime.timezone.utc)}]
```

And running a simple execution:
```python
>>> import pymssqlutils as sql
>>> result = sql.execute(
      "INSERT INTO mytable VALUES (1, 'test')",
      server="MySQLServer"
    )
```

### Specifying Connection
There are two ways of specifying the connection parameters to the SQL Server:
1. Passing the required parameters
   ([see pymssql docs](https://pymssql.readthedocs.io/en/stable/ref/pymssql.html#pymssql.connect))
   to `query` or `execute` like in the quickstart example above.
   All extra `**kwargs` passed to these methods are passed on to the `pymssql.connection()`.
2. Specify the connection parameters in the environment like the example below, this is the recommended way
   (however, any parameters given explicitly to a method will take priority).
   
```python
import os
import pymssqlutils as sql

os.environ["MSSQL_SERVER"] = "sqlserver.mycompany.com"
os.environ["MSSQL_USER"] = "my_login"
os.environ["MSSQL_PASSWORD"] = "my_password123"

result = sql.execute("INSERT INTO mytable VALUES (%s, %s)", (1, "test"))
```

There is a helper method to set this in code, see `set_connection_details` below.

### Executing SQL
#### Query

The `query` method executes a SQL Operation which does not commit the transaction & returns the result.

```python
query(
    operation: str,
    parameters: SQLParameters = None,
    raise_errors: bool = True,
    **kwargs,
) -> DatabaseResult:
```

Parameters:
 * `operation (str)`: the SQL operation to execute.
 * `parameters (SQLParameters)`: parameters to substitute into the operation,
   these can be a single value, tuple or dictionary.
 * `raise_errors (bool)`: whether to raise exceptions or to return the error information with the result.
 * Any extra kwargs are passed to _pymssql's_ `connect` method.

Returns a `DatabaseResult` class, see documentation below.

#### Execute

The `execute` method executes a SQL Operation which commits the transaction
& optionally returns the result (not by default).

```python
execute(
    operations: Union[str, List[str]],
    parameters: Union[SQLParameters, List[SQLParameters]] = None,
    batch_size: int = None,
    fetch: bool = False,
    raise_errors: bool = True,
    **kwargs,
) -> DatabaseResult:
```

Parameters:
 * `operations (Union[str, List[str]])`: the SQL Operation/s to execute.
 * `parameters (Union[SQLParameters, List[SQLParameters]])`: parameters to substitute into the operation/s,
   these can be a single value, tuple or dictionary OR this can be a list of any of the previous.
 * `batch_size (int)`: if specified concatenates the operations together according to the batch_size,
   this can vastly increase performance if executing many statements.
   Raises an error if set to True and both operations and parameters are singular.
 * `fetch (bool)`: if True returns the result from the LAST execution, by default false.  
 * `raise_errors (bool)`: whether to raise exceptions or to return the error information with the result.
 * Any extra kwargs are passed to _pymssql's_ `connect` method.

Returns a `DatabaseResult` class, see documentation below.

There are two ways of using this function:

Passing in a single operation (`str`) to operations:
* If parameters is singular, this calls `pymssql.execute()` and executes a single operation
* If parameters is plural, this calls `pymssql.execute_many()` and executes one execution per parameter set

Passing in multiple operations (`List[str]`) to operations:
* If parameters is None, this calls `pymssql.execute_many()` and executes one execution per operation
* If parameters is the same length as operations, this calls `pymssql.execute()` multiple times
  and executes one execution per operation.

Optionally `batch_size` can be specified to use string concatenation to batch the operations, this can
provide significant performance gains if executing 100+ small operations. This is similar to `fast_executemany`
found in the `pyodbc` package. A value of 500-1000 is a good default.

### DatabaseResult Class

One big difference between this library and _pymssql_ is that here
`execute` and `query` return an instance of the `DatabaseResult` class.

This class holds the returned data, if there is any, and provides
various useful attributes and methods to work with the result.

#### Attributes & Properties
 * `ok`: True if the execution did not error, else False. Only useful if using `raise_errors = False`,
   see below section on Error Handling.
 * `error`: Populated by the error raised during execution (if applicable). Only useful if using `raise_errors = False`.
 * `fetch`: True if results from the execution were fetched (e.g. if using `query`), else False.
 * `commit`: True if the execution was committed (i.e. if using `execute`), else False.
 * `columns`: A list of the column names in the dataset returned from the execution (if applicable)
 * `data`: The dataset returned from the execution (if applicable), this is a list of dictionaries.
 * `raw_data`: The dataset returned from the execution (if applicable), this is a list of tuples.
 * `set_count`: Returns the count of result sets that the execution returned, as an integer.

#### Methods
 * `to_dataframe`: (requires Pandas to be installed), returns the dataset as a DataFrame object.
   All args and kwargs are parsed to the DataFrame constructor.
 * `to_json`: returns the dataset as a json serialized string using the `orjson` library, make sure this 
   optional dependency is installed by running `pip install --upgrade pymssql-utils[json]`.
   Note that this will fail if your data contains `bytes` type values. By default, this method returns a string, but
   pass `as_bytes = True` to return a byte string. Specify `with_columns = True` to include the column names
   (rows as dictionaries instead of tuples).
 * `write_error_to_logger`: writes the error information to the library's logger, optionally pass a `name` parameter
   to allow you to easier indentify the query in the logging output.
 * `raise_error`: raises a `pymssqlutils.DatabaseError` from the underlying `pymssql` error,
   optionally pass a `name` parameter to allow you to easier indentify the query in the error output.
 * `next_set`: changes the class to return the data and metadata (columns etc) of the next result set. Returns True
   if there was a next set to move to, otherwise returns False and doesn't do anything.
 * `previous_set`: changes the class to return the data and metadata (columns etc) of the previous result set. Returns True
   if there was a previous set to move to, otherwise returns False and doesn't do anything.


### Error handling

Both `query` & `execute` take `raise_errors` as a parameter, which is by default `True`. This means that by default
_pymssql-utils_ will allow _pymssql_ to raise errors as normal.

Passing `raise_errors` as `False` will pass any errors onto the `DatabaseResult` class, which allows you
to handle errors gracefully using the `DatabaseResult` class (see above), e.g.:

```python
import pymssqlutils as db

result = db.query("Bad Operation", raise_errors=False)

if not result.ok: # result.ok will be False due to error
    
    # write the error to logging output
    result.write_error_to_logger('An optional query identifier to aid logging')
    
    # the error is stored under the error attribute
    error = result.error 
   
    # can always re-raise the error
    result.raise_error('Query Identifier')
```

This can be useful in situations where you do not want the error to propogate, e.g. if querying the database
as part of an API response.

### Utility Functions
#### set_connection_details

The `set_connection_details` method is a helper function which will set the value of
the relevant environment variable for the connection kwargs given.

Warning: this function has program wide side effects and will overwrite any
previously set connection details in the environment; therefore its usage is only recommended
in single script projects/notebooks. All the connections details will also be visible in your code.

The preferred method in production scenarios is to set the environment variables directly.

```python
def set_connection_details(
    server: str = None,
    database: str = None,
    user: str = None,
    password: str = None
) -> None:
```

Parameters:
* `server (str)`: the network address of the SQL server to connect to, sets 'MSSQL_SERVER' in the environment.
* `database (str)`: the default database to use on the SQL server, sets 'MSSQL_DATABASE' in the environment.
* `user (str)`: the user to authenticate against the SQL server with, sets 'MSSQL_USER' in the environment
* `password (str)`: the password to authenticate against the SQL server with, sets 'MSSQL_PASSWORD' in the environment

#### substitute_parameters

The `substitute_parameters` method does the same parameter substitution as `query` and `execute`, but returns the
substituted operation instead of executing it. This allows you to see the actual operation being run
against the database and is useful for debugging and logging.

```python
substitute_parameters(
    operation: str,
    parameters: SQLParameters
) -> str:
```

Parameters:
* `operation (str)`: The SQL operation requiring substitution.
* `parameters (SQLParameters)`: The parameters to substitute in.

Returns the parameter substituted SQL operation as a string.

Example:

```python3
>>> substitute_parameters("SELECT %s Col1, %s Col2", ("Hello", 1.23))
"SELECT N'Hello' Col1, 1.23 Col2"
```

#### to_sql_list

The `to_sql_list` method converts a Python iterable to a string form of the SQL equivalent list. This is useful
when creating dynamic SQL operations using the 'IN' operator.

```python
to_sql_list(
    listlike: Iterable[SQLParameter]
) -> str:
```

Parameters:
* `listlike (Iterable[SQLParameter])`: The iterable of SQLParameter to transform

Returns the SQL equivalent list as a string

Examples:

```python3
>>> to_sql_list([1, 'hello', datetime.now()])
"(1, N'hello', N'2021-03-22T10:56:27.981173')"
```

```python3
>>> my_ids = [1, 10, 21]
>>> f"SELECT * FROM MyTable WHERE Id IN {to_sql_list(my_ids)}"
'SELECT * FROM MyTable WHERE Id IN (1, 10, 21)'
```

#### model_to_values

The `model_to_values` method converts a Python mapping (e.g. dictionary of Pydantic model) to the SQL equivalent
values string. This is useful when creating dynamic SQL operations using the 'INSERT' statement.

```python3
model_to_values(
    model: Any,
    prepend: List[Tuple[str, str]] = None,
    append: List[Tuple[str, str]] = None,
) -> str:
```

Parameters:
* `model (Any)`: A mapping to transform, i.e. a dictionary or an object that has the __dict__ method implemented,
  with string keys and SQLParameter values.
* `prepend (List[Tuple[str, str]])`: prepend a variable number of columns to the beginning of the values statement.
* `append (List[Tuple[str, str]])`: append a variable number of columns to the end of the values statement.

Returns a string of the form: `([attr1], [attr2], ...) VALUES (val1, val2, ...)`.

Warning: prepended and appended columns are not parameter substituted,
this can leave your code open to SQL injection attacks.

Example:

```python3
>>> my_data = {'value': 1.56, 'insertDate': datetime.now()}

>>> model_to_values(my_data, prepend=[('ForeignId', '@Id')])
"([ForeignId], [value], [insertDate]) VALUES (@Id, 1.56, N'2021-03-22T13:58:33.758740')"

>>> f"INSERT IN MyTable {model_to_values(my_data, prepend=[('foreignId', '@Id')])}"
"INSERT IN MyTable ([foreignId], [value], [insertDate]) VALUES (@Id, 1.56, N'2021-03-22T13:58:33.758740')"
```

## Notes
### Type Parsing

_pymssql-utils_ parses SQL types to their native python types regardless of the environment.
This ensures consistent behaviour across various systems, see the table below for a comparison.

|                   | Windows         |            | Ubuntu          |            |
|-------------------|-----------------|------------|-----------------|------------|
| *SQL DataType*    | *pymssql-utils* | *pymssql*  | *pymssql-utils* | *pymssql*  |
| Date              | date            | date       | date            | str        |
| Binary            | bytes           | bytes      | bytes           | bytes      |
| Time1             | time            | time       | time            | str        |
| Time2             | time            | time       | time            | str        |
| Time3             | time            | time       | time            | str        |
| Time4             | time            | time       | time            | str        |
| Time5             | time            | time       | time            | str        |
| Time6             | time            | time       | time            | str        |
| Time7             | time            | time       | time            | str        |
| Small DateTime    | datetime        | datetime   | datetime        | datetime   |
| Datetime          | datetime        | datetime   | datetime        | datetime   |
| Datetime2         | datetime        | datetime   | datetime        | str        |
| DatetimeOffset0   | datetime        | bytes      | datetime        | str        |
| DatetimeOffset1   | datetime        | bytes      | datetime        | str        |
| DatetimeOffset2   | datetime        | bytes      | datetime        | str        |
| UniqueIdentifier  | str             | bytes      | str             | ???        |

## Testing

Install pytest to run non-integration tests via `pytest .`,
these tests mock the cursor results allowing the library to test locally.

To test against an MSSQL instance install `pytest-dotenv`.
Then create a `.env` file with `"TEST_ON_DATABASE"` set as a truthy value, as well as any
connection environemt variables for the MSSQL server.
These tests will then be run (not-skipped), e.g. `pytest . --envfile .test.env`

### Why _pymssql_ when Microsoft officially recommends _pyodbc_ (opinion)?

There are other minor reasons someone might prefer _pymssql_, e.g.:
1. _pymssql_ supports for MSSQL specific data types such as `Datetimeoffset`.
2. _pymssql's_ parameter subsitution is done client-side improving debug/logging ability.
3. _pymssql's_ drivers are easily installed, meaning that your code 'just works' in 
   more environments without extra steps, [e.g. this](https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server?view=sql-server-ver15)!

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.2] - 2022-08-03
### Changed
- `DatabaseResult.to_dataframe` does not take *args anymore (this would have thrown an error anyway).
- `DatabaseResult.to_dataframe` now throws a ValueError if data is passed as a kwarg.
### Fixed
- Fixed type hints (and added missing type hints). This library now passes mypy strict check.

## [0.4.1] - 2022-08-02
### Added
- Added a `py.typed` file so that mypy etc will treat this library as being typed.

## [0.4.0] - 2022-07-20
### Changed
- `DatabaseResult.to_dataframe()` will now return an empty dataframe with the correct
  column names if the result has column information but no rows. Warning: Pandas will
  not be able to infer the correct dtypes from these columns.
- Loosened version restrictions on `Pandas` and `ORJSON` dependencies to be any version number.


## [0.3.0] - 2022-01-21
### Added
- Added support for multiple result sets, you can move between result sets using the new
  `next_set()` and `previous_set()` methods on `DatabaseResult`. `DatabaseResult` class
  also has a new attribute `set_count` to tell you how many result sets the query
  returned.
- `DatabaseResult.to_json()` now has an optional `with_columns` parameter.
- Added some missing doc strings.


## [0.2.0] - 2022-01-21
### Added
- Added some more tests, coverage now at 92%.
### Changed
- BREAKING: DatabaseResult's `source_types`, `columns`, `raw_data` and `data`,
  attributes/properties now cannot be `None`, in order to decrease type ambiguity. To 
  facilitate this change, they will raise a ValueError if they are called when the 
  DatabaseResult instance errored, or fetch was false.
- The SQLParameter type hint (used in various places) is now just an alias for Any.
- Miscellaneous refactoring and type hinting/guarding throughout the library. The 
  library now passes a basic `pyright` check with no errors. 


## [0.1.5] - 2021-09-21
### Fixed 
- Handle specific length byte strings raising exceptions when parsing database result, [closes #15](https://github.com/invokermain/pymssql-utils/issues/15).

## [0.1.4] - 2021-09-14
### Fixed 
- SQL Type `Uniqueidentifier` is now handled correctly and returned as a string, [closes #13](https://github.com/invokermain/pymssql-utils/issues/13).

## [0.1.3] - 2021-08-16
### Fixed 
- Fixed incorrect type hint for `DatabaseResult.data`.

### Changed
- Data parsing now raises warning when handling an unrecognised type
and returns identity function mapping instead of raising an Exception.
- This package now requires "pymssql>=2.1.4" instead of "pymssql>=2". This was implicit anyway as 2.1.4
is the minimum version that supports Python 3.7.

### Added
- Added a TDS Protocol version warning for users using version 7.2 or below, this is only checked once.
- Added some more type hints to `DatabaseResult`.

## [0.1.2] - 2021-07-14
### Fixed 
- Fix an issue where the returned result would be limited to 10,000 rows, #11.

## [0.1.1] - 2021-07-08
### Changed
- Relaxed Python version syntax (effectively the same versions, just a less explicit description).

## [0.1.0] - 2021-07-08
### Removed
- Dropped support for Python 3.6, this might still work, but it is not guaranteed or tested.

### Changed
- Improved intended type inference using `Cursor.description` information.
- Optimized performance by only checking the first non-null item in a column.
- Optimized memory usage by fetching and cleaning results in batches of 10k.
- Moved project and pipelines to use Poetry.
- Use flake8, isort and black for linting.
- Removed unnecessary dateutil dependency.

### Fixed
- Fixed an edge case where strings with a mix of non-datelike and datelike formats would be parsed as mixed data types instead of just string.

tests/conftest.py
from contextlib import contextmanager

import pymssql
import pytest

from tests.helpers import MockConnection


@pytest.fixture()
def mock_pymssql_connect(monkeypatch):
    @contextmanager
    def mock_connect(as_dict, **kwargs):
        yield MockConnection(as_dict=as_dict)

    monkeypatch.setattr(pymssql, "connect", mock_connect)
    monkeypatch.setenv("MSSQL_SERVER", "test_server")


pymssqlutils/methods.py
import logging
import os
import warnings
from contextlib import contextmanager
from itertools import zip_longest
from typing import Any, Dict, Iterable, List, Optional, Tuple, Union, cast

import pymssql as sql
from pymssql import Connection

from .databaseresult import DatabaseResult
from .helpers import SQLParameter, SQLParameters

logger = logging.getLogger(__name__)

TDS_PROTOCOL_CHECKED = False


def substitute_parameters(operation: str, parameters: SQLParameters) -> str:
    """
    This function returns the SQL operation that would be executed on the server (i.e.
    after parsing and substituting of parameters). Useful for logging & debugging.

    :param operation: The SQL operation requiring substitution
    :param parameters: The parameters to substitute in
    :return: The parameter substituted SQL operation as a string
    """
    if isinstance(parameters, tuple):
        parameters = tuple(
            item.isoformat() if hasattr(item, "isoformat") else item
            for item in parameters
        )
    elif isinstance(parameters, dict):
        parameters = {
            key: item.isoformat() if hasattr(item, "isoformat") else item
            for key, item in parameters.items()
        }
    else:
        if hasattr(parameters, "isoformat"):
            parameters = parameters.isoformat()
        elif parameters is None:
            parameters = (None,)

    return cast(
        str, sql._mssql.substitute_params(operation, parameters).decode("UTF-8")
    )


def set_connection_details(
    server: Optional[str] = None,
    database: Optional[str] = None,
    user: Optional[str] = None,
    password: Optional[str] = None,
) -> None:
    """
    Sets the relevant environment variable for each passed parameter.

    :param server: the network address of the SQL server to connect to,
                   sets 'MSSQL_SERVER' in the environment
    :param database: the default database to use on the SQL server,
                     sets 'MSSQL_DATABASE' in the environment
    :param user: the user to authenticate against the SQL server with,
                 sets 'MSSQL_USER' in the environment
    :param password: the password to authenticate against the SQL server with,
                     sets 'MSSQL_PASSWORD' in the environment
    :return:
    """
    if server:
        os.environ["MSSQL_SERVER"] = server
    if database:
        os.environ["MSSQL_DATABASE"] = database
    if user:
        os.environ["MSSQL_USER"] = user
    if password:
        os.environ["MSSQL_PASSWORD"] = password


def _with_conn_details(kwargs: Dict[str, Optional[str]]) -> Dict[str, Optional[str]]:
    out = kwargs.copy()

    out["server"] = kwargs.get("server", os.environ.get("MSSQL_SERVER"))
    out["database"] = kwargs.get("database", os.environ.get("MSSQL_DATABASE"))
    out["user"] = kwargs.get("user", os.environ.get("MSSQL_USER"))
    out["password"] = kwargs.get("password", os.environ.get("MSSQL_PASSWORD"))

    if not out["server"]:
        raise ValueError(
            "`server` must be passed as a parameter or `MSSQL_SERVER` "
            "must be set in the environment."
        )

    return out


def query(
    operation: str,
    parameters: SQLParameters = None,
    raise_errors: bool = True,
    **kwargs: Optional[str],
) -> DatabaseResult:
    """
    Execute a SQL Operation which DOES NOT COMMIT the transaction & returns the result.

    **kwargs are passed through to the pymssql.connect() method.

    :param operation: the SQL Operation to execute
    :type operation: str
    :param parameters: parameters to substitute into the operation.
    :type parameters: SQLParameters
    :param raise_errors: if True raises errors, else DatabaseResult class will
                         contain the error details
    :type raise_errors: bool, optional
    :return: a DatabaseResult class.
    :rtype: DatabaseResult
    """
    try:
        return _execute(
            [operation],
            [parameters] if parameters else None,
            commit=False,
            fetch=True,
            **_with_conn_details(kwargs),
        )
    except sql.Error as err:
        if raise_errors:
            raise err
        return DatabaseResult(ok=False, fetch=True, commit=False, error=err)


def execute(
    operations: Union[str, List[str]],
    parameters: Union[SQLParameters, List[SQLParameters]] = None,
    batch_size: Optional[int] = None,
    fetch: bool = False,
    raise_errors: bool = True,
    **kwargs: Optional[str],
) -> DatabaseResult:
    """
    Used for a SQL Operation/s which COMMIT the transaction.
    There are two ways of using this function:

    Passing in a single operation (str) to operations:
        * If parameters is singular, this calls pymssql.execute()
          and executes a single operation
        * If parameters is plural, this calls pymssql.execute_many()
          and executes one execution per parameter set

    Passing in multiple operations (List[str]) to operations:
        * If parameters is None, this calls pymssql.execute_many()
          and executes one execution per operation
        * If parameters is the same length as operations, this calls pymssql.execute()
          multiple times and executes one execution per operation.

    Optionally batch_size can be specified to use string concatenation to batch
    the operations, this can provide significant performance gains if executing 100+
    small operations. This is similar to fast_executemany found in pyodbc package. A
    value of 500-1000 is a good default.

    Fetch can be set to True, this will only return the last result set.

    :param operations: the SQL Operation/s to execute. If this is a list then
                       parameters needs to be None or a list of the same length.
    :type operations: Union[str, List[str]]
    :param parameters: parameters to substitute into the operation. These can be a
                       single value, tuple or dictionary. If operations is a list
                       this parameter needs to either be None or a list of the
                       same length.
    :type parameters: Union[SQLParameters, List[SQLParameters]], optional
    :param batch_size: If specified concatenate the operations together
                       according to the batch_size, raises an error if set to True
                       and both operations and parameters are singular
    :type batch_size: int, optional
    :param fetch: return the LAST result of the execution
    :type fetch: bool, optional
    :param raise_errors: if True raises errors, else DatabaseResult class will
                         contain the error details
    :type raise_errors: bool, optional
    :return: a DatabaseResult class
    :rtype: DatabaseResult
    """
    if isinstance(operations, str):
        operations = [operations]
    if parameters is not None and not isinstance(parameters, list):
        parameters = [parameters]

    singular_operations = len(operations) == 1
    singular_parameters = parameters and len(parameters) == 1

    # validate
    if batch_size is not None:
        if singular_operations and singular_parameters:
            raise ValueError(
                "batch_size cannot be used if both operations "
                "and parameters are singular"
            )
        if batch_size <= 0:
            raise ValueError("batch_size cannot be negative")

    if parameters is not None:
        if not (singular_parameters or singular_operations) and len(operations) != len(
            parameters
        ):
            raise ValueError(
                "parameters must be the same length as "
                "operations if they are both lists"
            )

    try:
        if batch_size:
            return _execute_batched(
                operations, parameters, batch_size, fetch, **_with_conn_details(kwargs)
            )
        return _execute(
            operations,
            parameters,
            commit=True,
            fetch=fetch,
            **_with_conn_details(kwargs),
        )
    except sql.Error as err:
        if raise_errors:
            raise err
        return DatabaseResult(
            ok=False,
            fetch=fetch,
            commit=True,
            error=err,
        )


@contextmanager
def _get_connection(**kwargs: Union[str, int, bool, None]) -> Connection:
    conn = sql.connect(**kwargs)
    global TDS_PROTOCOL_CHECKED
    if not TDS_PROTOCOL_CHECKED:
        tds_major, tds_minor = conn._conn.tds_version_tuple
        if tds_major < 7 or tds_minor < 3:
            message = (
                f"Your connection is trying to use TDS Protocol {tds_major}.{tds_minor}"
                ", this protocol does not support time, date, "
                "datetime2 or datetimeoffset. "
                "It is strongly suggested you upgrade pymssql to the latest version."
            )
            warnings.warn(message, RuntimeWarning)
        TDS_PROTOCOL_CHECKED = True
    yield conn


def _execute(
    operations: List[str],
    parameters: Optional[List[SQLParameters]] = None,
    commit: bool = False,
    fetch: bool = False,
    **kwargs: Optional[str],
) -> DatabaseResult:
    """
    This is an internal method, you should call execute() instead
    """
    with _get_connection(**kwargs) as cnxn:
        with cnxn.cursor() as cur:
            if parameters:
                fillvalue = (
                    parameters[-1]
                    if len(parameters) < len(operations)
                    else operations[-1]
                )
                for operation, parameter_set in zip_longest(
                    operations, parameters, fillvalue=fillvalue
                ):
                    statement = substitute_parameters(
                        operation,  # type: ignore
                        parameter_set,
                    )
                    cur.execute(statement)
            else:
                for operation in operations:
                    cur.execute(operation)

            result = DatabaseResult(ok=True, fetch=fetch, commit=commit, cursor=cur)

        if commit:
            cnxn.commit()

    return result


def _execute_batched(
    operations: List[str],
    parameters: Optional[List[SQLParameters]] = None,
    batch_size: int = 1000,
    fetch: bool = False,
    **kwargs: Optional[str],
) -> DatabaseResult:
    """
    This is an internal method, you should call execute() instead
    """
    if parameters:
        fillvalue = (
            parameters[-1] if len(parameters) < len(operations) else operations[-1]
        )
        batched = [
            "\n;".join(
                substitute_parameters(operation, parameter_set)  # type: ignore
                for operation, parameter_set in zip_longest(
                    operations[i : i + batch_size],
                    parameters[i : i + batch_size],
                    fillvalue=fillvalue,
                )
            )
            for i in range(0, max(len(parameters), len(operations)), batch_size)
        ]
    else:
        batched = [
            "\n;".join(operation for operation in operations[i : i + batch_size])
            for i in range(0, len(operations), batch_size)
        ]

    with _get_connection(**_with_conn_details(kwargs)) as cnxn:
        with cnxn.cursor() as cur:
            for batch in batched:
                cur.execute(batch)
            result = DatabaseResult(ok=True, fetch=fetch, commit=True, cursor=cur)
        cnxn.commit()
    return result


def to_sql_list(listlike: Iterable[SQLParameter]) -> str:
    """
    Transforms an iterable to a SQL list string.
    Intended to be used with the SQL 'in' operator when building dynamic SQL queries.

    e.g. [1, 2, 3] -> '(1, 2, 3)'

    :param listlike: The iterable of SQLParameter to transform
    :return: str
    """
    out_str = ", ".join(substitute_parameters("%s", x) for x in listlike)
    return f"({out_str})"


def model_to_values(
    model: Any,
    prepend: Optional[List[Tuple[str, str]]] = None,
    append: Optional[List[Tuple[str, str]]] = None,
) -> str:
    """
    Transforms a Dict or Mapping into a string of the form:
    '([attr1], [attr2], ...) VALUES (val1, val2, ...)'.
    Intended to be used when creating dynamic SQL INSERT statements.

    Prepend and append can be used to add a SQL column with a static or variable value
    at the beginning or end of the values list.

    e.g. passing prepend = [('prependedColumn', '@prependedColumn')] would return a
    string of the form:
    '([prependedColumn], [attr1], ...) VALUES (@prependedColumn, val1, ...)'

    Note: prepended and appended columns are not parameter substituted,
          this can leave your code open to sql injection.

    :param model: a Dictionary or anything that implements the __dict__ method
                  (e.g. a Pydantic Model)
    :param prepend: prepend a variable number of columns to the beginning of the
                    values statement.
    :param append: append a variable number of columns to the end of the
                   values statement.
    :return: str
    """
    keys = [x[0] for x in prepend] if prepend else []
    prepend_values = [x[1] for x in prepend] if prepend else []
    append_values = [x[1] for x in append] if append else []

    if isinstance(model, dict):
        keys.extend(model.keys())
        values = list(model.values())
    else:
        keys.extend(model.__dict__.keys())
        values = list(model.__dict__.values())

    if append:
        keys.extend(x[0] for x in append)

    values_as_str = ", ".join(
        [
            *map(str, prepend_values),
            # keep (x,) due to: https://github.com/pymssql/pymssql/issues/696
            *[substitute_parameters("%s", (x,)) for x in values],
            *map(str, append_values),
        ]
    )

    column_names = "(" + ", ".join(f"[{key}]" for key in keys) + ")"
    return f"{column_names} VALUES ({values_as_str})"
